### PR TITLE
[fastrtps] Fix the macro definition in thread

### PR DESCRIPTION
**Pull request by jimwang118**
Fix [vcpkg testing MSVC](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=8171383&view=results) issues.
`D:\buildtrees\fastrtps\src\v2.7.0-02c5279eff.clean\include\fastrtps/utils/TimedMutex.hpp(142): error C2065: '_Thrd_success': undeclared identifier`
Starting from msvc 19.38.32926.95, the macro `_Thrd_success` is modified to `_Thrd_result::_Success` in thread, thus causing the above error.
We judge the `msvc` version in `cmakelists.txt`. If the version is greater than or equal to 19.38.32926.95, then modify the macro only used in `fastrtps` from `_Thrd_success` to _Thrd_result::_Success.
A issue [3783](https://github.com/eProsima/Fast-DDS/issues/3783) has been submitted upstream.

- [X] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [ ] ~~SHA512s are updated for each updated download~~
- [ ] ~~The "supports" clause reflects platforms that may be fixed by this new version~~
- [ ] ~~Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.~~
- [ ] ~~Any patches that are no longer applied are deleted from the port's directory.~~
- [X] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [X] Only one version is added to each modified port's versions file.

### CODE REVIEW

**Review by LilyWangLL — DISMISSED** [review]

File: ports/fastrtps/fix_thread.patch [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
++###############################################################################
++# Replace _Thrd_success with _Thrd_result::_Success. As a workround of the
++# unreleased version of MSVC, it will be deleted after release.
++###############################################################################
++if(CMAKE_CXX_COMPILER_ID MATCHES "MSVC")
++    if(CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL "19.38.32926.95")
++        file(READ "${PROJECT_SOURCE_DIR}/include/fastrtps/utils/TimedMutex.hpp" _contents)
++        string(REPLACE "_Thrd_success" "_Thrd_result::_Success" _contents "${_contents}")
++        file(WRITE "${PROJECT_SOURCE_DIR}/include/fastrtps/utils/TimedMutex.hpp" "${_contents}")
++    endif()
```
Comment (dg0yt) [inline ports/fastrtps/fix_thread.patch:19]: IMO this isn't an acceptable way to modify the source code  even for vcpkg:
- It breaks `--editable` mode.
- It creates concurrent access in parallel configure.

Instead add a compile definition, and add an additional `#ifdef` if needed, and fixup the *installed*  headers if needed.

**Review by LilyWangLL — DISMISSED** [review]

**Comment by BillyONeal** [comment]
It would be better to use the cast solution suggested by @StephanTLavavej in https://github.com/eProsima/Fast-DDS/issues/3783#issuecomment-1664943297 .

File: ports/fastrtps/fix_thread.patch [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+ #if defined(_WIN32)
+ class TimedMutex
+ {
++    // On MSVC 19.38.32926.95 `_Thrd_success` was changed into `_Thrd_result::_Success`.
++    // See https://github.com/eProsima/Fast-DDS/issues/3783
++    // See https://github.com/microsoft/STL/pull/3897
++#if defined(_MSC_FULL_VER) && _MSC_FULL_VER >= 193832926
++    using _Thrd_success = _Thrd_result::_Success;
++#endif
+ public:
```
Comment (JavierMatosD) [inline ports/fastrtps/fix_thread.patch:15]: I don't think this is the right solution. I believe the right solution is to follow @StephanTLavavej suggestion of `static_cast<int>` the value and comparing to 0 https://github.com/eProsima/Fast-DDS/issues/3783#issuecomment-1664943297

**Comment by LilyWangLL** [comment]
After applied this patch, `fastrtps` still build failed with the following error:
```
C:\PROGRA~1\MICROS~3\2022\ENTERP~1\VC\Tools\MSVC\1436~1.325\bin\HostX64\x64\cl.exe   /TP -DASIO_DISABLE_VISIBILITY -DASIO_STANDALONE -DBOOST_ASIO_STANDALONE -DFASTCDR_DYN_LINK -DFASTRTPS_DYN_LINK -DFASTRTPS_SOURCE -DFOONATHAN_MEMORY=1 -DFOONATHAN_MEMORY_VERSION_MAJOR=0 -DFOONATHAN_MEMORY_VERSION_MINOR=7 -DFOONATHAN_MEMORY_VERSION_PATCH=3 -DNOMINMAX -DSHM_TRANSPORT_BUILTIN -DSQLITE_WIN32_GETVERSIONEX=0 -DTINYXML2_DEBUG -DTINYXML2_IMPORT -D_ENABLE_ATOMIC_ALIGNMENT_FIX -D_WIN32_WINNT=0x0601 -Dfastrtps_EXPORTS -IE:\Lily\fastrp\buildtrees\fastrtps\src\v2.7.0-831b33fc2a.clean\thirdparty\nlohmann-json -IE:\Lily\fastrp\buildtrees\fastrtps\src\v2.7.0-831b33fc2a.clean\thirdparty\filewatch -IE:\Lily\fastrp\buildtrees\fastrtps\src\v2.7.0-831b33fc2a.clean\include -IE:\Lily\fastrp\buildtrees\fastrtps\x64-windows-dbg\include -IE:\Lily\fastrp\buildtrees\fastrtps\x64-windows-dbg\include\fastrtps -IE:\Lily\fastrp\buildtrees\fastrtps\src\v2.7.0-831b33fc2a.clean\src\cpp -IE:\Lily\fastrp\buildtrees\fastrtps\src\v2.7.0-831b33fc2a.clean\thirdparty\boost\include -IE:\Lily\fastrp\buildtrees\fastrtps\src\v2.7.0-831b33fc2a.clean\thirdparty\taocpp-pegtl -external:IE:\Lily\fastrp\installed\x64-windows\include -external:W0 /nologo /DWIN32 /D_WINDOWS /W4 /utf-8 /GR /EHsc /MP  /D_DEBUG /MDd /Z7 /Ob0 /Od /RTC1  -MDd /w34101 /w34189 /w34555 /w34715 /w35038 /showIncludes /Fosrc\cpp\CMakeFiles\fastrtps.dir\rtps\resources\ResourceEvent.cpp.obj /Fdsrc\cpp\CMakeFiles\fastrtps.dir\ /FS -c E:\Lily\fastrp\buildtrees\fastrtps\src\v2.7.0-831b33fc2a.clean\src\cpp\rtps\resources\ResourceEvent.cpp
E:\Lily\fastrp\buildtrees\fastrtps\src\v2.7.0-831b33fc2a.clean\include\fastrtps/utils/TimedMutex.hpp(46): error C3861: '_Mtx_init': identifier not found
E:\Lily\fastrp\buildtrees\fastrtps\src\v2.7.0-831b33fc2a.clean\include\fastrtps/utils/TimedMutex.hpp(56): error C3861: '_Mtx_destroy': identifier not found
E:\Lily\fastrp\buildtrees\fastrtps\src\v2.7.0-831b33fc2a.clean\include\fastrtps/utils/TimedMutex.hpp(117): error C3861: '_Mtx_init': identifier not found
E:\Lily\fastrp\buildtrees\fastrtps\src\v2.7.0-831b33fc2a.clean\include\fastrtps/utils/TimedMutex.hpp(127): error C3861: '_Mtx_destroy': identifier not found
E:\Lily\fastrp\buildtrees\fastrtps\src\v2.7.0-831b33fc2a.clean\include\fastrtps/utils/TimedMutex.hpp(142): error C2677: binary '==': no global operator found which takes type '_Thrd_result' (or there is no acceptable conversion)
```

**Pull request closed without merging**